sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -197,24 +197,41 @@ private ITypeSymbol HandlePropertyOrField(IdentifierNameSyntax identifier, ISymb
 
             return methodSymbol.IsExtensionMethod
                 ? semanticModel.GetTypeInfo(invokedOn).ConvertedType
-                : FindOriginatingSymbol(invocationSymbol);
+                : FindOriginatingSymbol(invocationSymbol, callSite);
         }
 
-        private INamedTypeSymbol FindOriginatingSymbol(ISymbol accessedMember)
+        private static INamedTypeSymbol FindOriginatingSymbol(ISymbol accessedMember, IAssemblySymbol usageSite)
         {
             if (accessedMember == null)
             {
                 return null;
             }
 
             var originatingInterface = accessedMember.GetInterfaceMember()?.ContainingType;
-            if (originatingInterface != null)
+            if (originatingInterface != null &&
+                IsNotInternalOrSameAssembly(originatingInterface))
             {
                 return originatingInterface;
             }
 
-            return SymbolHelper.GetOverriddenMember(accessedMember)?.ContainingType
-                ?? accessedMember.ContainingType;
+            var originatingType = SymbolHelper.GetOverriddenMember(accessedMember)?.ContainingType;
+            if (originatingType != null &&
+                IsNotInternalOrSameAssembly(originatingType))
+            {
+                return originatingType;
+            }
+
+            return accessedMember.ContainingType;
+
+            // Do not suggest internal types that are declared in an assembly different than
+            // the one that's declaring the parameter. Such types should not be suggested at
+            // all if there is no InternalsVisibleTo attribute present in the compilation.
+            // Since the check for the attribute must be done in CompilationEnd thus making
+            // the rule unusable in Visual Studio, we will not suggest such classes and will
+            // generate some False Negatives.
+            bool IsNotInternalOrSameAssembly(INamedTypeSymbol namedTypeSymbol) =>
+                namedTypeSymbol.ContainingAssembly == usageSite ||
+                namedTypeSymbol.GetEffectiveAccessibility() != Accessibility.Internal;

```

Comment:
That's the important change - we ignore internal types that originate from another assembly than the one that is declaring the analyzed parameter